### ISSUE TITLE
MIDIInputMap and MIDIOutputMap are bizarre

### ISSUE THREAD

**Issue by domenic** [WebAudio/web-midi-api#91]
The keys, entries, and values functions are weird. What kind of function do they return? It doesn't seem to be used in any of the examples. Why can't they just return iterables instead of taking callback functions? How do you envision the example

``` js
for (output in outputs) {
      var opt = document.createElement("option");
      opt.text = input.name;
      document.getElementById("inputportselector").add(opt);
    }
```

working? Are you going to add enumerable properties for every output?

Why not just follow the standard ECMAScript pattern, and have keys, entries, and values return iterables of strings, [string, MIDIOutput] pairs, and MIDIOutputs, respectively? You'd also want to add an `@@iterator` aliasing `values`, so that people could use `for`-`of` (_not_ `for`-`in`) over the MIDIOutputMap instance itself, instead of having to call values. (Although if you are trying to emulate `Map`, aliasing `entries` might be better and less surprising to consumers.)

**Comment by cwilso** [comment]
I think I'm trying to do approximately what you're suggesting; however, I've found nearly zero guidance on how to do this, and end up struggling with ReSpec the rest of the time.

The goal is simply to have a collection of inputs (and outputs) that can be iterated through, or searched by an opaque identifier.  Iterated order is not a primary identifier (i.e. the collection is dynamic, and can change quickly), so arrays are not good patterns.

**Comment by domenic** [comment]
OK cool. Yeah, I don't think WebIDL has the capacity to do this, and I can't imagine ReSpec has any ability to handle it either.

**Comment by toyoshim** [comment]
I modified them to be more compatible with ES6.
https://github.com/WebAudio/web-midi-api/pull/98

**Comment by cwilso** [comment]
Domenic, could you take another look and see if Takashi's changes address this issue fully?

**Comment by domenic** [comment]
It looks pretty good. Only a couple things:
- You do not have a property that makes the maps actually iterable. In ES6 this would be done via `MIDIOutputMap.prototype[Symbol.iterator] = MIDIOutputMap.prototype.entries`, but I am not sure how to do so in WebIDL. I think @heycam is still planning to add iteration features to WebIDL?
- `forEach` should take a `thisArg`, and its arguments should be `(value, key, map)`, not `(key, value)`.

Also I noticed sometimes you say `key` and sometimes you say `id`, which is a bit confusing.

**Comment by heycam** [comment]
Web IDL already has the [iterator](http://heycam.github.io/webidl/#idl-iterators) keyword, but it is still defined to create a property named `iterator` rather than using `Symbol.iterator`. That will get fixed some time soon.

**Comment by yutakahirano** [comment]
WebIDL has MapClass attribute now. Can we use it?
http://heycam.github.io/webidl/#MapClass

**Comment by heycam** [comment]
My feeling is that we shouldn't be trying to kind-of emulate standard Map objects.  Tab is proposing to do so (well, for Sets) in the Font Loading API spec but I am trying to convince him otherwise.  People didn't seem to like ArrayClass, as it lets you create objects that feel kind of like Array objects but which aren't quite the same, and I think the same arguments apply to MapClass/SetClass.  I don't know that I have a good suggestion on what should be done instead though.  I'm interested to hear what Dominic thinks.  (I assume the ES spec still doesn't allow us to subclass Map while enforcing additional requirements like the argument to `get` and `has` being stringified.)

**Comment by yutakahirano** [comment]
@domenic , do you have an opinion?

@heycam , If it is not good to define a map-like class, what do you think about MIDIInputMap and MIDIOutputMap? They are map-like objects.
Can returning a true map be a solution, though we have to create and return a map for each call because ES6 Map is mutable?

**Comment by domenic** [comment]
My opinion is that it's a great idea to emulate standard interfaces. MapClass and SetClass and ArrayClass are bad since they actually do prototypal inheritance, but just reusing method names so that people share a vocabulary is a good idea. What's the alternative---making up your own method names for the same operations? That would be silly.

Overall feedback on the current design is that it looks fairly solid. It needs an iterator property, but WebIDL is missing that technology for now; perhaps it could be defined in prose, i.e. something like "Additionally, `MIDIInputMap.prototype[Symbol.iterator] = MIDIInputMap.prototype.entries`." Also, I think the forEach design is a bit off: maps do `map.forEach((value, key, map) => { ... }, thisP)`.

**Comment by jussi-kalliokoski** [comment]
I agree with Domenic; I don't think prototypal inheritance is necessary here, but implementing the standard interface is very convenient, although not from the spec point of view, since WebIDL doesn't have good support for this (yet).

**Comment by yutakahirano** [comment]
Thanks! Then I will implement as currently specified.
But the map-like interface will be used in many places and it would be great if WebIDL has the terminology for it, I think.

**Comment by yutakahirano** [comment]
I agree with Domenic in terms of making MIDIInputMap itself iterable. Then, you can use
for (var input of midiAccess.inputs) { }
that is equivalen to
for (var input of midiAccess.inputs.entries()) {}
.

**Comment by domenic** [comment]
The WebIDL map-like discussion is happening at https://www.w3.org/Bugs/Public/show_bug.cgi?id=26183. If you guys have feedback on a good way for spec authors to provide some kind of abstract key-value model that `[MapLikeIterable]` can refer to, that would be the most helpful thing for moving that forward.

**Comment by yutakahirano** [comment]
We now have [(readonly) maplike](http://heycam.github.io/webidl/#idl-maplike) in WebIDL spec.

**Comment by toyoshim** [comment]
this could be closed or merged to #115
